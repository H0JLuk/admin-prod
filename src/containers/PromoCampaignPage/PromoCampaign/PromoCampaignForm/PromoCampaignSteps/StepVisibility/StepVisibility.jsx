import React, { useCallback } from 'react';
import { Button } from 'antd';
import PromoCampaignVisibilitySettingInput from '../../../PromoCampaignVisibilitySettingInput/PromoCampaignVisibilitySettingInput';
import PromoCampaignVisibilitySetting from '../../../../PromoCampaign/PromoCampaignVisibilitySetting/PromoCampaignVisibilitySetting';

import cross from '../../../../../../static/images/cross.svg';

import styles from './StepVisibility.module.css';

const TITLE_OF_STEP = 'Настройки видимости';
const ADD_SETTING = 'Добавить настройку';
const EMPTY_SETTING = { location: null, salePoint: null, visible: false, errors: {} };

const StepVisibility = ({
    visibilitySettings = [],
    onChangeState,
    onDeleteState,
    viewMode,
    isCopy,
    copyVisibilitySettings,
}) => {

    const onChange = (val, idx, input, error = {}) => {
        const hasErrors = !!Object.keys(error).length;
        onChangeState(val, idx, input, hasErrors);

        if (input === 'location') {
            onChangeState({}, idx, 'errors', hasErrors);
        }
    };

    const addVS = useCallback(() => {
        onChangeState({ ...EMPTY_SETTING, id: Date.now() }, visibilitySettings.length);
    }, [visibilitySettings, onChangeState]);

    return viewMode && !isCopy && !copyVisibilitySettings ? (
        <PromoCampaignVisibilitySetting
            searchAndSortMode={ false }
            hideHeader
            addNewByModal
        />
    ) : (
        <div className={ styles.containerStep }>
            <div className={ styles.titleStep }>{ TITLE_OF_STEP }</div>
            { visibilitySettings.map((setting, idx) => (
                <div key={ setting.id } className={ styles.content }>
                    { idx !== 0 && (
                        <img
                            src={ cross }
                            alt="delete"
                            className={ styles.deleteBlock }
                            onClick={ () => onDeleteState(idx, !!Object.keys(setting.errors).length) }
                        />
                    ) }
                    <PromoCampaignVisibilitySettingInput
                        error={ setting.errors }
                        onLocationChange={ (loc) => onChange(loc, idx, 'location', setting.errors) }
                        onSalePointChange={ (salePoint) => onChange(salePoint, idx, 'salePoint', setting.errors) }
                        visibility={ setting.visible }
                        onVisibilityChange={ () => onChange(!setting.visible, idx, 'visible') }
                        location={ setting.location }
                        salePoint={ setting.salePoint }
                    />
                </div>
            )) }
            <Button
                type="primary"
                onClick={ addVS }
            >
                { ADD_SETTING }
            </Button>
        </div>
    );
};

export default StepVisibility;