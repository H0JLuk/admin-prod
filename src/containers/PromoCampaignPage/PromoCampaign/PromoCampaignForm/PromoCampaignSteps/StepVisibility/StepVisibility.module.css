.containerStep {
    padding: 0 15px;
}

.titleStep {
    font-size: 24px;
    font-weight: bold;
    margin-bottom: 20px;
}

.content {
    position: relative;
    margin-bottom: 15px;
}

.deleteBlock {
    position: absolute;
    right: 5px;
    top: 5px;
    cursor: pointer;
}