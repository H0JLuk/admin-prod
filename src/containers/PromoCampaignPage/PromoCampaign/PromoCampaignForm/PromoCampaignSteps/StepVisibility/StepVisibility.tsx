import React, { useCallback } from 'react';
import { Button } from 'antd';
import { PromoCampaignFormVisibilitySettingCreateDto } from '../../types';
import PromoCampaignVisibilitySettingInput from '../../../PromoCampaignVisibilitySettingInput';
import PromoCampaignVisibilitySetting from '../../../PromoCampaignVisibilitySetting';
import { LocationDto, SalePointDto } from '@types';

import cross from '@imgs/cross.svg';

import styles from './StepVisibility.module.css';

type OnChangeStateFunc = (
    val: LocationDto | boolean | SalePointDto | null | emptySettingsType | unknown,
    idx: number,
    input?: string,
    hasErrors?: boolean
) => void;

type emptySettingsType = { id: number; } & typeof EMPTY_SETTING;

export type StepVisibilityProps = {
    visibilitySettings: PromoCampaignFormVisibilitySettingCreateDto[];
    onChangeState: OnChangeStateFunc;
    onDeleteState: (idx: number, vis: boolean) => void;
    viewMode: boolean;
    isCopy: boolean | undefined;
    copyVisibilitySettings: PromoCampaignFormVisibilitySettingCreateDto[];
};

const TITLE_OF_STEP = 'Настройки видимости';
const ADD_SETTING = 'Добавить настройку';
const EMPTY_SETTING = { location: null, salePoint: null, visible: false, errors: {} };

const StepVisibility: React.FC<StepVisibilityProps> = ({
    visibilitySettings = [],
    onChangeState,
    onDeleteState,
    viewMode,
    isCopy,
    copyVisibilitySettings,
}) => {

    const onChange = (val: LocationDto | boolean | SalePointDto | null, idx: number, input: string, error = {}) => {
        const hasErrors = !!Object.keys(error).length;
        onChangeState(val, idx, input, hasErrors);

        if (input === 'location') {
            onChangeState({}, idx, 'errors', hasErrors);
        }
    };

    const addVS = useCallback(() => {
        onChangeState({ ...EMPTY_SETTING, id: Date.now() }, visibilitySettings.length);
    }, [visibilitySettings, onChangeState]);

    return viewMode && !isCopy && !copyVisibilitySettings ? (
        <PromoCampaignVisibilitySetting
            searchAndSortMode={false}
            hideHeader
            addNewByModal
        />
    ) : (
        <div className={styles.containerStep}>
            <div className={styles.titleStep}>{TITLE_OF_STEP}</div>
            {visibilitySettings.map((setting, idx) => (
                <div key={setting.id} className={styles.content}>
                    {idx !== 0 && (
                        <img
                            src={cross}
                            alt="delete"
                            className={styles.deleteBlock}
                            onClick={() => onDeleteState(idx, !!Object.keys(setting.errors).length)}
                        />
                    )}
                    <PromoCampaignVisibilitySettingInput
                        error={setting.errors}
                        onLocationChange={(loc) => onChange(loc, idx, 'location', setting.errors)}
                        onSalePointChange={(salePoint) => onChange(salePoint, idx, 'salePoint', setting.errors)}
                        visibility={setting.visible}
                        onVisibilityChange={() => onChange(!setting.visible, idx, 'visible')}
                        location={setting.location}
                        salePoint={setting.salePoint}
                    />
                </div>
            ))}
            <Button
                type="primary"
                onClick={addVS}
            >
                {ADD_SETTING}
            </Button>
        </div>
    );
};

export default StepVisibility;
