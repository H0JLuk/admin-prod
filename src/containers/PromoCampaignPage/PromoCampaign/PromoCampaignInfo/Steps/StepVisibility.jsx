import React from 'react';
import PromoCampaignVisibilitySetting from '../../../PromoCampaign/PromoCampaignVisibilitySetting/PromoCampaignVisibilitySetting';

const StepVisibility = () => (
    <PromoCampaignVisibilitySetting hideHeader searchAndSortMode={ false } addNewByModal />
);

export default StepVisibility;