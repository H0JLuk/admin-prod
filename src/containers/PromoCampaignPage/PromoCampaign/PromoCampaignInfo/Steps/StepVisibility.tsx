import React from 'react';
import PromoCampaignVisibilitySetting from '../../PromoCampaignVisibilitySetting';

const StepVisibility = () => (
    <PromoCampaignVisibilitySetting hideHeader searchAndSortMode={false} addNewByModal />
);

export default StepVisibility;
