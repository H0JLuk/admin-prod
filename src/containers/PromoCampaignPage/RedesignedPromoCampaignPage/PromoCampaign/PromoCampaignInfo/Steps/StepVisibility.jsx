import React from 'react';
import PromoCampaignVisibilitySetting from '../../../../PromoCampaign/PromoCampaignVisibilitySetting/PromoCampaignVisibilitySetting';

const StepVisibility = () => (
    <PromoCampaignVisibilitySetting hideHeader={ true } searchAndSortMod = { false } />
);

export default StepVisibility;