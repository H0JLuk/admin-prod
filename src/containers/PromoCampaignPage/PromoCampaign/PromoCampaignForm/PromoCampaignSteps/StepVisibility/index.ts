import StepVisibility from './StepVisibility';

export type { StepVisibilityProps } from './StepVisibility';

export default StepVisibility;
