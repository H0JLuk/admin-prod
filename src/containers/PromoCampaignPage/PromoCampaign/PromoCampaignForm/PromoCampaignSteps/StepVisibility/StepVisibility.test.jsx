import { shallow } from 'enzyme';
import React from 'react';
import StepVisibility from './StepVisibility';

const mockHistoryPush = jest.fn();
jest.mock('react-router-dom', () => ({
    useHistory: () => ({
        push: mockHistoryPush,
        location: jest.fn()
    }),
    useRouteMatch: () => ({ path: 'test' }),
    useParams: () => ({ promoCampaignId: 1064 }),
    useLocation: () => ({
        search: {},
        state: 'test state'
    })
}));

jest.mock(
    '../../../PromoCampaignVisibilitySetting/PromoCampaignVisibilitySetting',
    () => () => <div>PromoCampaignVisibilitySetting</div>
);

jest.mock('', () => ({
    getPromoCampaignVisibilitySettings: () => ({
        visibilitySettings: [],
        pageNo: 0,
        totalElements: 0
    })
}));

describe('<StepVisibility /> tests', () => {
    const props = {
        visibilitySettings: [
            {
                errors: {},
                id: 1619691826728,
                location: null,
                salePoint: null,
                visible: false
            },
        ],
        onChangeState: jest.fn(),
        onDeleteState: jest.fn(),
        viewMode: true,
        isCopy: undefined,
        copyVisibilitySettings: undefined,
    };

    it('StepVisibility snapshot', () => {
        const StepVisibilityItem = shallow(<StepVisibility { ...props } />);
        expect(StepVisibilityItem.html()).toMatchSnapshot();
    });

    it('should click switch, button call onChangeState func', () => {
        const newProps = { ...props, copyVisibilitySettings: true };
        const wrapper = shallow(<StepVisibility { ...newProps } />);
        wrapper.find('PromoCampaignVisibilitySettingInput').simulate('visibilityChange');
        expect(props.onChangeState).toBeCalled();
        wrapper.find('Button').simulate('click');
        expect(props.onChangeState).toBeCalled();
    });

    it('should call onLocationChange and onSalePointChange funcs', () => {
        const newProps = { ...props, copyVisibilitySettings: true };
        const wrapper = shallow(<StepVisibility { ...newProps } />);
        wrapper.find('PromoCampaignVisibilitySettingInput').simulate('locationChange');
        expect(props.onChangeState).toBeCalled();
        wrapper.find('PromoCampaignVisibilitySettingInput').simulate('salePointChange');
        expect(props.onChangeState).toBeCalled();
    });

    it('should call onDelete prop func', () => {
        const newProps = {
            ...props,
            copyVisibilitySettings: true,
            visibilitySettings: [
                ...props.visibilitySettings,
                {
                    errors: {},
                    id: 1619691826728,
                    location: null,
                    salePoint: null,
                    visible: false,
                },
            ],
        };

        const wrapper = shallow(<StepVisibility { ...newProps } />);
        wrapper.find('.deleteBlock').simulate('click');
        expect(props.onDeleteState).toBeCalled();
    });
});
