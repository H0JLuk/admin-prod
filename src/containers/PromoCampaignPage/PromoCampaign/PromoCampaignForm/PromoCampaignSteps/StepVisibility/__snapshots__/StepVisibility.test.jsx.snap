// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StepVisibility /> tests StepVisibility snapshot 1`] = `"<div>PromoCampaignVisibilitySetting</div>"`;
